### PR TITLE
feat(datetime): add clear button

### LINKED ISSUE #17482: ionic-datetime does not support clearing the value set
# Bug Report

**Ionic version:**

[x] **4.x**

**Current behavior:**

ionic-datetime does not support clearing the value once it is set

**Expected behavior:**

Allow clearing the datetime value set

**Steps to reproduce:**

Here is the ionic-datetime demo - see https://ionicframework.com/docs/api/datetime
It is not possible to clear/reset the value

Please see the issue for Ionic 3 https://github.com/ionic-team/ionic/issues/11961

### PR DESCRIPTION
**Pull request by willmartindev**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: resolves #17482 

## What is the new behavior?

- A "Clear" button now can be optionally rendered next to the default "Done" and "Cancel" buttons by using the `showClearButton` prop.
- The text of the button can be modified using the `clearText` prop.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/datetime/datetime.tsx [hunk 1]
```diff
@@ -414,6 +429,16 @@ export class Datetime implements ComponentInterface {
     }
   }
 
+  /**
+   * Resets the internal state of the datetime
```
Comment (liamdebeasi) [inline core/src/components/datetime/datetime.tsx:433]: What does the internal state get reset to?

File: core/src/components/datetime/usage/angular.md [hunk 2]
```diff
@@ -40,6 +40,9 @@
   <div slot="title">My Custom Title</div>
 </ion-datetime>
 
+<!-- Clear button -->
+<ion-datetime showClearButton></ion-datetime>
```
Comment (liamdebeasi) [inline core/src/components/datetime/usage/angular.md:44]: Angular templates in strict mode might complain unless we pass in a boolean.

```suggestion
<ion-datetime [showClearButton]="true"></ion-datetime>
```

File: core/src/components/datetime/usage/react.md [hunk 3]
```diff
@@ -66,6 +66,9 @@ export const DateTimeExamples: React.FC = () => {
       <IonDatetime>
         <div slot="title">My Custom Title</div>
       </IonDatetime>
+
+      {/* Clear button */}
+      <IonDatetime showClearButton></IonDatetime>
```
Comment (liamdebeasi) [inline core/src/components/datetime/usage/react.md:71]:
```suggestion
      <IonDatetime showClearButton={true}></IonDatetime>
```

File: core/src/components/datetime/usage/stencil.md [hunk 4]
```diff
@@ -64,6 +64,9 @@ export class DatetimeExample {
       <ion-datetime>
         <div slot="title">My Custom Title</div>
       </ion-datetime>,
+
+      {/* Clear button */}
+      <ion-datetime showClearButton></ion-datetime>
```
Comment (liamdebeasi) [inline core/src/components/datetime/usage/stencil.md:69]:
```suggestion
      <ion-datetime showClearButton={true}></ion-datetime>  
```

File: core/src/components/datetime/usage/vue.md [hunk 5]
```diff
@@ -40,6 +40,9 @@
   <ion-datetime>
     <div slot="title">My Custom Title</div>
   </ion-datetime>
+
+  <!-- Clear button -->
+  <ion-datetime show-clear-button></ion-datetime>
```
Comment (liamdebeasi) [inline core/src/components/datetime/usage/vue.md:45]:
```suggestion
  <ion-datetime :show-clear-button="true"></ion-datetime>
```

File: core/src/components/datetime/datetime.tsx [hunk 6]
```diff
@@ -389,14 +404,15 @@ export class Datetime implements ComponentInterface {
...
-   * te component to the provided date.
+   * Resets the internal state of the datetime to `internalState`
+   * and optionally updates the value if `updateValue` is true.
+   * Passing a valid ISO-8601 string will reset the state of
+   * the component to the provided date.
    */
   @Method()
-  async reset(value?: string) {
-    this.processValue(value);
+  async reset(internalState?: string, updateValue = false) {
```
Comment (liamdebeasi) [inline core/src/components/datetime/datetime.tsx:413]: What if developers want to set the `value` to `undefined` but reset the internal state to a particular date?

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

File: core/src/components/datetime/datetime.tsx (hunk 6 above)
Comment (willmartindev) [reply core/src/components/datetime/datetime.tsx:413]: Hmm, `reset` could take two values instead?

`reset(internalState?: string, componentValue?: string)`

**Review by liamdebeasi — APPROVED** [review]
Great job!

**Pull request merged**